File: DSU/dsu.java
class DSU {
  private int[] par;
  private int[] rnk;
  private int[] sz;
  private int c;
  public DSU(int n) {
      par = new int[n+1];
      rnk = new int[n+1];
      sz = new int[n+1];
      c=n;
      sz[0]=1;
      for (int i = 1; i <= n; ++i) {
        par[i] = i;
        sz[i] = 1;
      }
  }
  public int find(int i) {
    return (par[i] == i ? i : (par[i] = find(par[i])));
  }
  public int get_size(int i){
    return sz[find(i)];
  }
  public boolean same(int i, int j) {
    return find(i) == find(j);
  }
  public void swap(int i, int j){
      int temp = i;
      i = j;
      j = temp;
  }
  public int count(){
      return c;
  }
  public Boolean merge(int i, int j) {
    if ((i = find(i)) == (j = find(j))) return false;
    else --c;
    if (rnk[i] > rnk[j]) swap(i, j);
    par[i] = j;
    sz[j] += sz[i];
    if (rnk[i] == rnk[j]) rnk[j]++;
    return true;
  }
};

File: DSU/dsu.py
class DSU :
    def __init__(self, n: int):
        self.par = list(range(n))
        self.rnk = [0] * (n)
        self.c = 0
        
    def find(self, i: int) -> int:
            if self.par[i] != i:
                self.par[i] = self.find(self.par[i])
            return self.par[i]
        
    def same(self, i: int, j: int):
            return self.find(i) == self.find(j)
        
    def count(self) ->int :
        return self.c
    
    def merge(self, i: int, j: int) :
            i = self.find(i)
            j = self.find(j)
            
            if i == j:
                return
            else:
                c -= 1
            if self.rnk[i] > self.rnk[j]:
                i, j = j, i
            self.par[i] = j
            if self.rnk[i] == self.rnk[j]:
                self.rnk[j] += 1
